### PR TITLE
Drop the consumer_manager module completely

### PR DESCRIPTION
**Pull request by dblia**
This PR implements the [Redefine Rafka scope](https://github.com/skroutz/rafka/blob/master/ddocs/design-rafka-rethinking.rst#redefine-rafka-scope) and [Drop redundant functionality](https://github.com/skroutz/rafka/blob/master/ddocs/design-rafka-rethinking.rst#drop-redundant-functionality) proposals of the [Rafka Rethinking](https://github.com/skroutz/rafka/blob/master/ddocs/design-rafka-rethinking.rst) design doc. For more details on the topic, please refer to those references.

A lot of major changes are introduced by this PR which are detailed in the respective commit messages. The most important ones are that the `Client` struct becomes aware of the server's `Context` as well as the Rafka configuration file. Since `Client` becomes the authoritative handler of `Consumer` objects, those two extra fields are essential to properly register and de-register a `Consumer` instance via the `Client` object.

Moreover, the `consumer` field is also added to the `Client`, which is a reference to the registered `Consumer` object. Also, a new function is introduced in `client` module, named `registerConsumer` which will be used to create a new `Consumer` instance upon request. The `Consumer` struct is extended with a new field containing the parent's cancel context (aka `Client`), which will allow us signal a `Consumer` instance about the Client's termination; a task which was performed until now by the `ConsumerManager` handler.

### CODE REVIEW

**Comment by ctrochalakis** [comment]
During `rafka` shutdown sequence, we waited for consumers to close gracefully (using the manager `wg`). This is now done in a non-blocking way and might result in non-committed offsets or other unexpected behavior.

**Comment by ctrochalakis** [comment]
Other that that, great work :) I think that practice has shown that having a single consumer per client is enough with the added benefit of a simpler overall design.

**Comment by dblia** [comment]
> During `rafka` shutdown sequence, we waited for consumers to close gracefully (using the manager `wg`). This is now done in a non-blocking way and might result in non-committed offsets or other unexpected behavior.

Indeed, the Server's shutdown flow has been modified while it shouldn't. I'll restore the original behavior shortly. Nice catch, thank you!

File: server.go [hunk 1]
```diff
@@ -490,7 +472,7 @@ func (s *Server) shutdown() {
 			return false
 		}
 
-		if c.consID != "" {
+		if c.consumer != nil {
```
Comment (agis) [inline server.go]: For the shake of encapsulation, it'd be better to add a method here, `Client.Consumer() *Consumer` and use that instead of accessing an unexported struct field.

File: client.go [hunk 2]
```diff
@@ -87,39 +90,70 @@ func (c *Client) SetID(id string) error {
...
+	go cons.Run(ctx)
+
+	return cons, nil
 }
 
-// fetchConsumer returns the registered Consumer for the current Client.
-func (c *Client) fetchConsumer() (*Consumer, error) {
-	if c.consID == "" {
-		return nil, errors.New("No consumer is registered for Client " + c.id)
+func (c *Client) Consumer(topics []string, cfg rdkafka.ConfigMap) (*Consumer, error) {
```
Comment (agis) [inline client.go]: While we're at it, we should add some documentation to this method. I also believe it should be renamed to `GetOrCreateConsumer` or something.

File: server.go [hunk 3]
```diff
@@ -501,9 +483,6 @@ func (s *Server) shutdown() {
 		return true
 	})
 
-	// wait for consumers to actually close
-	s.managerWg.Wait()
```
Comment (ctrochalakis) [inline server.go]: Also note that `managerWg.Wait()` **together** with `manager.StopAcceptingConsumers()` ensures that no new consumer is **created** during the consumer shutdown.

File: client.go (hunk 2 above)
Comment (agis) [reply client.go]: Actually, it should be `NewConsumer`.

File: server.go (hunk 1 above)
Comment (dblia) [reply server.go]: Ack! I'll update the PR accordingly.

File: server.go (hunk 3 above)
Comment (dblia) [reply server.go]: Right! However, I'm oriented to handling this scenario a bit differently. Currently, during a server's shutdown, we accept connections from new `Clients` and apparently new `Producers` while we don't accept new `Consumers` as you already mentioned.

Since there's no valid reason to create new clients and/or producers which will be closed some seconds/milliseconds later, we should modify this logic to first stop accepting new Client connections, before we start closing the registered consumers and producers.

Comment (agis) [reply server.go]: > Currently, during a server's shutdown, we accept connections from new `Clients` and apparently new `Producers` while we don't accept new `Consumers` as you already mentioned.

Correct. This is intentional, see https://github.com/skroutz/rafka/commit/af07cbf443789e59f2b9fbf495370dd7ad811139.

File: server.go (hunk 1 above)
Comment (dblia) [reply server.go]: Unfortunately this change is not that straightforward. Since the `ConsumerID` contains all the topics in which the `Consumer` is subscribed to, we can't simply call the `Consumer` upon a `RPUSH` request for a multi-topic `Consumer`, since offsets commits operation only run per single topic.

I agree with your rational, however I'd suggest to handle it separately in a subsequent PR.

File: client.go (hunk 2 above)
Comment (dblia) [reply client.go]: See, my comment [here](https://github.com/skroutz/rafka/pull/84#discussion_r328969782). Let's handle this scenario in another PR.

**Comment by dblia** [comment]
There's an non-handled case in `server: Revisit shutdown flow ` commit, which will be addresses shortly.

File: client.go [hunk 4]
```diff
@@ -100,29 +97,24 @@ func (c *Client) Consumer(topics []string, cfg rdkafka.ConfigMap) (*Consumer, er
...
-			}
-		}
+	if c.consID == "" {
+		c.consID = cid
 	}
 
-	c.consumers[cid] = true
-	for _, topic := range topics {
-		c.consByTopic[topic] = cid
+	if cid != c.consID {
```
Comment (agis) [inline client.go]: I think the following version is easier to follow:

```golang
if cid != "" || cid != c.consID {
  return nil, fmt.Errorf("Client '%s' has already Consumer '%s' registered.", c.id, c.consID)
}

c.consID = cid

// ...
```

WDYT?

File: client.go [hunk 5]
```diff
@@ -100,29 +97,24 @@ func (c *Client) Consumer(topics []string, cfg rdkafka.ConfigMap) (*Consumer, er
...
 	}
 
 	return c.consManager.GetOrCreate(cid, c.consGID, topics, cfg)
 }
 
-func (c *Client) ConsumerByTopic(topic string) (*Consumer, error) {
-	consumerID, ok := c.consByTopic[topic]
-	if !ok {
-		return nil, fmt.Errorf("No consumer for topic %s", topic)
+// fetchConsumer returns the registered Consumer for the current Client.
```
Comment (agis) [inline client.go]: Consider rephrasing to:

> fetchConsumer returns the registered Consumer of c, or nil if none exists.

File: client.go [hunk 6]
```diff
@@ -100,29 +97,24 @@ func (c *Client) Consumer(topics []string, cfg rdkafka.ConfigMap) (*Consumer, er
...
+func (c *Client) fetchConsumer() (*Consumer, error) {
+	if c.consID == "" {
+		return nil, errors.New("No consumer is registered for Client " + c.id)
```
Comment (agis) [inline client.go]: I don't think we should return an error in this case. It means a consumer does not exist, so I wouldn't consider this an exceptional case.

File: server.go [hunk 7]
```diff
@@ -490,7 +490,7 @@ func (s *Server) shutdown() {
 			return false
 		}
 
-		if len(c.consumers) > 0 {
+		if c.consID != "" {
```
Comment (agis) [inline server.go]: Can't we reuse `fetchConsumer()` here? If so, we should make it exported (`FetchConsumer`).

File: client.go [hunk 8]
```diff
@@ -87,39 +90,69 @@ func (c *Client) SetID(id string) error {
...
 
-	return c.consManager.GetOrCreate(cid, c.consGID, topics, cfg)
+	return cons, nil
 }
 
-// fetchConsumer returns the registered Consumer for the current Client.
-func (c *Client) fetchConsumer() (*Consumer, error) {
-	if c.consID == "" {
-		return nil, errors.New("No consumer is registered for Client " + c.id)
+// Consumer method will either create a new consumer via the registerConsumer method, or it will
```
Comment (agis) [inline client.go]: nit: I'd remove mentioning `registerConsumer` as it's an implementation detail:

> Consumer returns the client's consumer. If one doesn't exist, it registers a new one and returns that instead.

Also, I think this should be named `GetOrCreateConsumer()`, which is more precise.

File: client.go [hunk 9]
```diff
@@ -87,39 +90,69 @@ func (c *Client) SetID(id string) error {
 	return nil
 }
 
-func (c *Client) Consumer(topics []string, cfg rdkafka.ConfigMap) (*Consumer, error) {
-	if !c.consReady {
-		return nil, errors.New("Connection not ready. Identify yourself using `CLIENT SETNAME` first")
-	}
+// registerConsumer registers a new Consumer denoted by cid.
+func (c *Client) registerConsumer(
+	cid ConsumerID, gid string, topics []string, cfg rdkafka.ConfigMap) (*Consumer, error) {
```
Comment (agis) [inline client.go]: As a sanity check, I think we should assert that gid is not an empty string. If it is, we should return an error.

File: client.go [hunk 10]
```diff
@@ -87,39 +90,69 @@ func (c *Client) SetID(id string) error {
...
+	cid := ConsumerID(fmt.Sprintf("%s|%s", c.id, strings.Join(topics, ",")))
+
+	if c.consumer == nil {
+		cons, err := c.registerConsumer(cid, c.consGID, topics, cfg)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to register Consumer '%s' for Client '%s'", cid, c.id)
+		}
+		c.consumer = cons
+	} else if cid != c.consumer.id {
+		return nil, fmt.Errorf("Client '%s' has already Consumer '%s' registered.", c.id, c.consumer.id)
```
Comment (agis) [inline client.go]: It'd be useful to also print `cid` here, for debugging purposes.

File: server.go [hunk 11]
```diff
@@ -112,7 +108,7 @@ func (s *Server) monitorHandler() {
 }
 
 func (s *Server) Handle(ctx context.Context, conn net.Conn) {
-	c := NewClient(conn, s.manager)
+	c := NewClient(ctx, conn, cfg)
```
Comment (agis) [inline server.go]: I'd consider making `NewClient` a method on `Server`. This way we wouldn't have to pass `cfg` everytime.

File: server.go [hunk 12]
```diff
@@ -466,6 +476,9 @@ Loop:
 // shutdown closes current clients and also stops accepting new clients in a
 // non-blocking manner
 func (s *Server) shutdown() {
+	// Stop accepting new consumers while a server shutdown operation is in-progress.
+	s.teardown = true
```
Comment (agis) [inline server.go]: There is a possible data race here. Since `s.teardown` may be accessed concurrently by `shutdown()` and `Handle()`, it should be guarded with a lock. In this particular case, see [`RWMutex`](https://golang.org/pkg/sync/#RWMutex).

File: docs/server-shutdown.rst [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+======================
+
+This document details the sequence of operations executed when Rafka server receives a shutdown
+signal. The steps executed by the `Server` are summarized in the following list:
+
+#. Receives the shutdown request
+#. Stops accepting new `Consumers`
+#. Closes all `Clients` with an active `Consumer` instance
+#. Waits for all `Consumers` to terminate
+#. Stops accepting new `Clients`
```
Comment (agis) [inline docs/server-shutdown.rst]: I think we should rephrase this to `Stops accepting new `Producers`, for clarity.

File: server.go (hunk 3 above)
Comment (0xdiba) [reply server.go]: Not directly related to the current PR but since it's been discussed, I think that having different shutdown policies complicates things and might lead to unflushed messages.
In the case we allow a producer to register itself and produce a high number of messages during our shutdown process we will end up [flushing a number of messages while closing](https://github.com/skroutz/rafka/blob/master/producer.go#L78) leaving a high probability of unflushed messages. Those unflushed messages won't be communicated to the client ( as is the case in rafka in general ) and due to the shutdown they won't event get reported since `STATS` will be overridden. That said, imo letting producers late-register might lead to unnoticed problems ( deficient monitoring is partly to blame here too )

Comment (agis) [reply server.go]: Isn't this case also possible given a "unified policy" flow? For example, assuming the flow:

1. stop accepting new clients
2. close consumers & producers
3. shutdown

In the above case, we can run in the same problem you mentioned, if a producer registers right before (1) and produces a large number of messages.

The possibility of unflushed messages will always be there IMHO and is irrelevant to rafka's shutdown process. That said, we should implement monitoring and handling of unflushed messages. Also #35 is something that should help.

Comment (0xdiba) [reply server.go]: Of course this can happen in any flow.
The difference is that now we allow more time for producers to register and the window for unflushed messages just gets bigger.

Regarding #35, for producers that need to make sure that all their messages are flushed [Producer.Flush](https://github.com/skroutz/rafka/blob/f9311550d7ec82bf7ddec7e5094d34912780bf60/producer.go#L70) can be used.

In any case, this is out of the current PR's scope and I was just expressing some concerns on the subject. We'll revisit this when it's time comes.

File: server.go (hunk 12 above)
Comment (0xdiba) [reply server.go]: Since `Handle` only reads the value to check if it should consume or not we should probably decide if we want any [performance regression](https://github.com/golang/go/issues/17973) that might get introduced by this mutex or we can live with having a consumer trying to consume after we initiate a `teardown`.

File: client.go (hunk 4 above)
Comment (dblia) [reply client.go]: This is an outdated commit; as mentioned in the respective commit message, `consID` was an intermediate step to allow a smoother drop of _ConsumerManager_.

File: client.go (hunk 5 above)
Comment (dblia) [reply client.go]: outdated, too!

File: client.go (hunk 6 above)
Comment (dblia) [reply client.go]: ditto! in the updated code path we don't raise an error in that case:

```
if c.consumer == nil {
        cons, err := c.registerConsumer(cid, c.consGID, topics, cfg)
	if err != nil {
		return nil, fmt.Errorf("Failed to register Consumer '%s' for Client '%s'", cid, c.id)
	}
	c.consumer = cons
} else if cid != c.consumer.id {
	return nil, fmt.Errorf("Client '%s' has already Consumer '%s' registered.", c.id, c.consumer.id)
}
```

File: server.go (hunk 7 above)
Comment (dblia) [reply server.go]: `Consumer` is the `fetchConsumer` alternative. However, is not straightforward to call that function without refactoring the code in this place.

File: client.go (hunk 8 above)
Comment (dblia) [reply client.go]: I don't really like exposing the implementation details to the function name. Overall, you request a Consumer object regardless of what we do in the background (aka register a new Consumer or returning an already registered Consumer).

So, I'd vote for leaving this as it for now.

Comment (agis) [reply client.go]: Fair enough.

File: client.go (hunk 9 above)
Comment (dblia) [reply client.go]: The `gid` attribute is set by the `SetID` function and there's no code-path to call `registerConsumer` without calling the `SetID` function first.

However, we could add an extra layer of protection here.

File: client.go (hunk 10 above)
Comment (dblia) [reply client.go]: ack.

File: server.go (hunk 11 above)
Comment (dblia) [reply server.go]: This requires yet another refactoring. Let's not include it in the current PR and keep the PR as contained as possible.

File: docs/server-shutdown.rst (hunk 13 above)
Comment (dblia) [reply docs/server-shutdown.rst]: It actually stops accepting new Clients not Producers in this step. You can't initialize a connection with Rafka after this step, so changing it to _Stop accepting new Producers_ would confuse a user because it would except its connection be accepted.

I'd say to leave this as is for now.

File: client.go [hunk 14]
```diff
@@ -90,48 +93,73 @@ func (c *Client) SetID(id string) error {
...
+	if err != nil {
+		return nil, err
+	}
 
-	c.consumers[cid] = true
-	for _, topic := range topics {
-		c.consByTopic[topic] = cid
+	// Extract the consumer name from the client id.
+	// We know by client.Consumer() that cid is in the form of `<group:name>|<topics>`
+	cidNoTopics := strings.Split(strings.Split(string(cid), "|")[0], ":")[1]
```
[the remark this answers is not in the document]
Comment (dblia) [reply client.go]: I agree, however this code path was already there so let's leave it as is for the current PR.

File: server.go (hunk 3 above)
Comment (dblia) [reply server.go]: Let's not mess things up! This is part of another PR/issue/etc.

File: client.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (0xdiba) [reply client.go]: :+1:

File: server.go (hunk 12 above)
Comment (dblia) [reply server.go]: Interesting case. We could benchmark the performance regression of maintaining a mutex in this variable and decide accordingly IMHO. I suggest to better handle this in a separate PR and create an issue to not forget it. What do you think @agis, @0xdiba ?

Comment (agis) [reply server.go]: Synchronization is needed here, otherwise it means consumers might be created *after* `shutdown()` is called, which can lead to all sorts of undefined behavior (since we're dealing with librdkafka under the hood). Best-case scenario, consumers may be outstanding and not finalized (commit offsets etc.) when rafka shuts down.

Regarding the referenced upstream issue, it won't affect us since the cost of calling RWLock/RWUnlock is amortized by the work `Consume()` is doing. It would be an issue if we had a hot-path calling RWLock very fast, but in rafka this is not an issue.

That said, if RWMutex's performance was a concern, we could use `sync.Atomic` or channels or something else.

Comment (dblia) [reply server.go]: Agreed, I'll proceed as you suggest.

**Comment by dblia** [comment]
Squashed all commits and forced-push for a last look.

File: server.go [hunk 15]
```diff
@@ -112,7 +116,8 @@ func (s *Server) monitorHandler() {
 }
 
 func (s *Server) Handle(ctx context.Context, conn net.Conn) {
-	c := NewClient(conn, s.manager)
+	c := NewClient(ctx, conn, cfg)
+	c.consWg = &s.consumersWg
```
Comment (0xdiba) [inline server.go]: I would prefer server and client having separate wait groups :) but we should address this in another iteration.

Comment (dblia) [reply server.go]: Sure, let's revisit this in another PR!

**Comment by agis** [comment]
:rocket: :clap: :tada:
